cloudinit/socket.py

```diff
@@ -0,0 +1,164 @@
+
+LOG = logging.getLogger(__name__)
+
+
+def sd_notify(message: str):
+    """Send a sd_notify message.
+
+    :param message: sd-notify message (must be valid ascii)
+    """
+    socket_path = os.environ.get("NOTIFY_SOCKET", "")
+
+    if not socket_path:
+        # not running under systemd, no-op
+        return
+
+    elif socket_path[0] == "@":
+        # abstract
+        socket_path.replace("@", "\0", 1)
+
+    # unix domain
+    elif socket_path[0] != "/":
+        raise OSError("Unsupported socket type")
+
+    with socket.socket(
+        socket.AF_UNIX, socket.SOCK_DGRAM | socket.SOCK_CLOEXEC
+    ) as sock:
+        LOG.info("Sending sd_notify(%s)", str(message))
+        sock.connect(socket_path)
+        sock.sendall(message.encode("ascii"))
+
+
+class SocketSync:
+    """A two way synchronization protocol over Unix domain sockets."""
+
+    def __init__(self, *names: str):
+        """Initialize a synchronization context.
+
+        1) Ensure that the socket directory exists.
+        2) Bind a socket for each stage.
+
+        Binding the sockets on initialization allows receipt of stage
+        "start" notifications prior to the cloud-init stage being ready to
+        start.
+
+        :param names: stage names, used as a unique identifiers
+        """
+        self.stage = ""
+        self.remote = ""
+        self.first_exception = ""
+        self.sockets = {
+            name: socket.socket(
+                socket.AF_UNIX, socket.SOCK_DGRAM | socket.SOCK_CLOEXEC
+            )
+            for name in names
+        }
+        # ensure the directory exists
+        os.makedirs(f"{DEFAULT_RUN_DIR}/share", mode=0o700, exist_ok=True)
+        # removing stale sockets and bind
+        for name, sock in self.sockets.items():
+            socket_path = f"{DEFAULT_RUN_DIR}/share/{name}.sock"
+            with suppress(FileNotFoundError):
+                os.remove(socket_path)
+            sock.bind(socket_path)
+
+    def __call__(self, stage: str):
+        """Set the stage before entering context.
+
+        This enables the context manager to be initialized separately from
+        each stage synchronization.
+
+        :param stage: the name of a stage to synchronize
+
+        Example:
+            sync = SocketSync("stage 1", "stage 2"):
+            with sync("stage 1"):
+                pass
+            with sync("stage 2"):
+                pass
+        """
+        if stage not in self.sockets:
+            raise ValueError(f"Invalid stage name: {stage}")
+        self.stage = stage
+        return self
+
+    def __enter__(self):
+        """Wait until a message has been received on this stage's socket.
+
+        Once the message has been received, enter the context.
+        """
+        if os.isatty(sys.stdin.fileno()):
+            LOG.info("Not syncing, stdin is a tty.")
+            return
+        sd_notify(
+            "STATUS=Waiting on external services to "
```

Comment:
I really like the supplemental Status here messages here. It gives a very clear view from systemctl status cloud-init.service what is ultimately blocking the run of cloud-init:
```
● cloud-init.service - Cloud-init: Single Process
     Loaded: loaded (/usr/lib/systemd/system/cloud-init.service; enabled; prese>
     Active: active (running) since Wed 2024-07-24 13:15:23 UTC; 1min 45s ago
 Invocation: 7a47a36e765b474c829f5419d6da84d1
   Main PID: 108 (cloud-init)
     Status: "Waiting on external services to complete (network stage)"
      Tasks: 3
```


Bikeshed: given that this point in time is actually waiting for the `start` message from the stage and the run of that entire stage can we change the message to
`Waiting on external services to run (network stage)`

Reply:
I think that we might be interpreting this differently. This message was intended to give the context that the network stage is waiting on external services to complete, which is what I think you are trying to change it to say. Perhaps the parenthesis are adding confusion?

How about:

"Waiting on external services to complete before starting the network stage."

I think that should eliminate confusion without including parenthesis.

Reply:
@blackboxsw let me know what you think of the wording in the latest commit

Reply:
+1 on your updated text `Waiting on external services to complete before starting the network stage.`. Yes, I was interpreting the message as saying the network stage had started and we are awaiting its completion.